R/sort_linter.R

```diff
@@ -0,0 +1,89 @@
+#' Require usage of `sort()` over `.[order(.)]`
+#'
+#' `sort()` is the dedicated option to sort a list or vector. It is more legible
+#' and around twice as fast as `.[order(.)]`, with the gap in performance
+#' growing with the vector size.
+#'
+#' @examples
+#' # will produce lints
+#' lint(
+#'   text = "x[order(x)]",
+#'   linters = sort_linter()
+#' )
+#'
+#' lint(
+#'   text = "x[order(x, decreasing = TRUE)]",
+#'   linters = sort_linter()
+#' )
+#'
+#' # okay
+#' lint(
+#'   text = "x[sample(order(x))]",
+#'   linters = sort_linter()
+#' )
+#'
+#' lint(
+#'   text = "y[order(x)]",
+#'   linters = sort_linter()
+#' )
+#'
+#' @evalRd rd_tags("sort_linter")
+#' @seealso [linters] for a complete list of linters available in lintr.
+#' @export
+sort_linter <- function() {
+  xpath <- "
+  //OP-LEFT-BRACKET
+    /following-sibling::expr[
+      expr[1][
+        SYMBOL_FUNCTION_CALL[text() = 'order']
+        and following-sibling::expr =
+          parent::expr
+            /parent::expr
+            /expr
+      ]
+    ]
+  "
+
+  Linter(function(source_expression) {
+    if (!is_lint_level(source_expression, "expression")) {
+      return(list())
+    }
+
+    xml <- source_expression$xml_parsed_content
+
+    bad_expr <- xml2::xml_find_all(xml, xpath)
+
+    var <- xml2::xml_text(xml2::xml_find_first(bad_expr, ".//SYMBOL"))
+
+    args <- ".//SYMBOL_SUB[text() = 'method' or
+                           text() = 'decreasing' or
+                           text() = 'na.last']"
+
+    arg_names <- xml2::xml_text(xml2::xml_find_all(bad_expr, args))
```

Comment:
I think this will run into trouble on expressions with `>1` lint, try an example like:

```r
c(
  x[order(x, decreasing = TRUE)],
  y[order(y, decreasing = TRUE, na.last = FALSE)]
)
```

because `xml_find_all()` can return any number of results per input, there's no way to may between input & output when `length(bad_expr) > 1`

Reply:
Yes, you're right and I don't know how to resolve this :disappointed:

Reply:
Could you add some failing tests that should pass if this is resolved?

I'll have a go at fixing it, but it would help a lot to start with a more tangible target.

Reply:
I've added a failing test in 266ddb65ec2ec90552f07b0f49f078fe0b48dc2b. Thanks!

Reply:
Should be fixed by [`64a88b4` (#1528)](https://github.com/r-lib/lintr/pull/1528/commits/64a88b49c24d72afcf84fd2e3ea758d9f7ad1b99).